ytypes/leafref.go

```diff
@@ -189,23 +189,13 @@ func dataNodesAtPath(ni *util.NodeInfo, path *gpb.Path, pathQueryNode *util.Path
 			if root.Parent == nil {
 				return nil, fmt.Errorf("no parent for leafref path at %v, with remaining path %s", ni.Schema.Path(), path)
 			}
-			if !util.IsCompressedSchema(root.Schema) && root.Parent.Schema.IsList() && util.IsValueMap(root.Parent.FieldValue) {
-				// If we are in an uncompressed schema, then we have one more level of the data tree than
-				// the YANG expects, since our data tree layout is:
-				// struct (parent container)
-				//  --> map (the list)
-				//  --> struct (the list member)
-				//
-				// In YANG, .. from the list member struct gets us to the parent container, but for us
-				// we have only reached the map. This means that we end up over-consuming the ".."s. To
-				// avoid this issue, if we are in an uncompressed schema, and in a list, and we find
-				// that we're looking at the map, we consume another level of the data tree. This gets
-				// us to the parent container with ".." as would be expected.
-				//
-				// This is NOT required for the compressed schema, because in this case, we have removed
-				// a level of the data tree. So the parent container in the above example will have been
-				// removed. This is enforced by ygen. In this case, we do want to consume the extra level
-				// of ..s in the list case, such that we do not end up under-consuming them.
+			if (root.Parent.Schema.IsList() && util.IsValueMap(root.Parent.FieldValue)) || (root.Parent.Schema.IsLeafList() && util.IsValueSlice(root.Parent.FieldValue)) {
```

Comment:
Ah, this is interesting -- as long as the current test cases pass then I'm good with this. I think the iteration around this code has been that confusion as to whether this is the case in a compressed schema vs. an uncompressed one.

Inspecting the code, I'm not quite sure without reloading full context how we avoid the issue that we are at `/foo/surrounding-container/list/element` and have a reference to `/foo/bar`. In this case `element` can have a leafref `../../../bar` but in the generated compressed structs we don't have the `surrounding-container`. I /think/ this is handled because we're iterating the schema and avoiding that level, but I just want to confirm that this case is still handled with this change.